internal/olm/operator/registry/operator_installer_test.go

```diff
@@ -16,22 +16,233 @@ package registry
 )
 
+type FakeCatalogCreator struct {
+	client crclient.Client
+}
+
+func (f FakeCatalogCreator) CreateCatalog(ctx context.Context, name string) (*v1alpha1.CatalogSource, error) {
+	cs := &v1alpha1.CatalogSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "MockCatalogSource",
+		},
+	}
+	if err := f.client.Create(ctx, cs); err != nil {
+		return nil, fmt.Errorf("Unable to create CatalogSource with fake client: %v", err)
+	}
+	return cs, nil
+}
+
 var _ = Describe("OperatorInstaller", func() {
-	Describe("InstallOperator", func() {
-		// TODO: fill this in once run bundle is done
+	Describe("NewOperatorInstaller", func() {
+		It("should create an OperatorInstaller", func() {
+			kubeconfigPath := os.Getenv(k8sutil.KubeConfigEnvVar)
+			cfg := &operator.Configuration{KubeconfigPath: kubeconfigPath}
+			// TODO(asmacdo) Is there anything I can test here that wouldnt be caught by the compiler?
+			// TODO(asmacdo)It doesnt even throw an err.
+			_ = NewOperatorInstaller(cfg)
+		})
+	})
+
+	Describe("createSubscription", func() {
+		var (
+			oi  *OperatorInstaller
+			sch *runtime.Scheme
+		)
+		BeforeEach(func() {
+			// Setup and fake client
+			kubeconfigPath := os.Getenv(k8sutil.KubeConfigEnvVar)
+			cfg := &operator.Configuration{KubeconfigPath: kubeconfigPath}
+			sch = runtime.NewScheme()
+			Expect(v1.AddToScheme(sch)).To(Succeed())
+			Expect(v1alpha1.AddToScheme(sch)).To(Succeed())
+			cfg.Client = fake.NewFakeClientWithScheme(sch)
+
+			// Create OperatorInstaller struct
+			oi = NewOperatorInstaller(cfg)
+			oi.CatalogCreator = FakeCatalogCreator{
+				client: cfg.Client,
+			}
+			oi.StartingCSV = "fakeName"
+			oi.cfg.Namespace = "fakeNS"
+		})
+
+		It("should create the subscription with the fake client", func() {
+			sub, err := oi.createSubscription(context.TODO(), "huzzah")
+			Expect(err).ToNot(HaveOccurred())
+
+			retSub := &v1alpha1.Subscription{}
+			subKey := types.NamespacedName{
+				Namespace: sub.GetNamespace(),
+				Name:      sub.GetName(),
+			}
+			err = oi.cfg.Client.Get(context.TODO(), subKey, retSub)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(retSub.GetName()).To(Equal(sub.GetName()))
+			Expect(retSub.GetNamespace()).To(Equal(sub.GetNamespace()))
+		})
+
+		It("should pass through any client errors (duplicate)", func() {
+
+			sub := newSubscription(oi.StartingCSV, oi.cfg.Namespace,
+				withPackageChannel(oi.PackageName, oi.Channel, oi.StartingCSV),
+				withCatalogSource("duplicate", oi.cfg.Namespace),
+				withInstallPlanApproval(v1alpha1.ApprovalManual))
+			oi.cfg.Client = fake.NewFakeClientWithScheme(sch, sub)
+
+			_, err := oi.createSubscription(context.TODO(), "duplicate")
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).Should(ContainSubstring("error creating subscription"))
+		})
+	})
+
+	// TODO(asmacdo) Since this uses the olm client can/should this be unit tested?
+	// Describe("getInstalledCSV", func() {
+	// 	})
+	// })
```

Comment:
Not in this PR. I think we need to refactor the code to make it testable. I don't like when we create clients there:
```
c, err := olmclient.NewClientForConfig(o.cfg.RESTConfig)
```
We can do the option pattern in this method:

```golang
type Option func(...)

func WithOLMClient(...)  {
}

func (o OperatorInstaller) getInstalledCSV(ctx context.Context, opts...lOption) (*v1alpha1.ClusterServiceVersion, error) {
...
}
```

Then in the test we can create a fake client and pass it in via the `WithOLMClient`. In the main line code, it can create the client like it does. We had to do something similar to the leader election code in operator-lib:
* https://github.com/operator-framework/operator-lib/blob/main/leader/leader.go#L91-L104
* https://github.com/operator-framework/operator-lib/blob/main/leader/leader.go#L53-L82

Reply:
@jmrodri I'll file an issue.

Reply:
https://github.com/operator-framework/operator-sdk/issues/4173